railties/lib/rails/application/default_middleware_stack.rb

```diff
@@ -18,7 +18,10 @@ def build_stack
           middleware.use ::Rack::Sendfile, config.action_dispatch.x_sendfile_header
 
           if config.serve_static_files
-            middleware.use ::ActionDispatch::Static, paths["public"].first, config.static_cache_control, config.static_index
+            headers = config.public_file_server.headers
+            headers['Cache-Control'] = config.static_cache_control if config.static_cache_control.present?
+
+            middleware.use ::ActionDispatch::Static, paths["public"].first, :not_set, config.static_index, headers: headers
```

Comment:
I needed to use `:not_set` explicitly due to this change: https://github.com/rails/rails/pull/20017. I think we should make the `index` arg a keyword argument so we don't have to care about the order of the args here. Alternatively, we can just merge the `static_index` config into the `public_file_server` namespace, which I actually prefer, but I'm open to suggestions.

cc/ @eliotsykes @jonatack